### PR TITLE
Add conformer w2v2 model architecture

### PR DESCRIPTION
**Pull request by carolineechen**
internal comparison tests: D40080919

follow up PR for pretrained models #2827

### CODE REVIEW

File: docs/source/refs.bib [hunk 1]
```diff
@@ -439,3 +439,16 @@ @article{coucke2018snips
   journal={arXiv preprint arXiv:1805.10190},
   year={2018}
 }
+@article{conformerssl,
```
Comment (carolineechen) [inline docs/source/refs.bib:442]: note: use icassp paper

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+from torchaudio.models.wav2vec2 import components
+
+
+class TimeReduction(Module):
+    """Extract features from input. Consists of time reduction and linear layer.
+
+    Args:
+        stride (int): number of frames to merge for the output frame
+        input_dim (int): input dimension of the tensor
+        output_dim (int): output dimension of the tensor
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:17]: the order of the arguments can be same as the one in `_get_conformer_feature_extractor` for consistency

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+    def forward(
+        self,
+        x: Tensor,
+        length: Optional[Tensor],
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        """
+        Args:
+            x (Tensor):
+                Input Tensor representing Mel Spectrogram output. shape ``[batch, frame, dim]``.
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:33]:
```suggestion
                Input Tensor representing Mel Spectrogram output. shape ``(batch, time, dim)``.
```

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+from torch import nn, Tensor
+from torch.nn import Module
+from torchaudio.models import Wav2Vec2Model
+from torchaudio.models.conformer import ConformerLayer
+from torchaudio.models.rnnt import _TimeReduction
+from torchaudio.models.wav2vec2 import components
+
+
+class TimeReduction(Module):
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:11]: would it be better to name it as `FeatureExtractor` or `ConformerFeatureExtractor`? want to hear thoughts of other folks @hwangjeff @mthrok @xiaohui-zhang

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 3 above)
Comment (nateanl) [reply torchaudio/prototype/models/conformer_wav2vec2.py:33]: is the feature mel spectrogram or log mel spectrogram?

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+    def __init__(self, stride: int, input_dim: int, output_dim: int):
+        super().__init__()
+        self.time_reduction_layer = _TimeReduction(stride=stride)
+        self.linear_layer = nn.Linear(input_dim * stride, output_dim)
+
+    def forward(
+        self,
+        x: Tensor,
+        length: Optional[Tensor],
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:28]: should `length` be `lengths` for consistency?

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        features: Tensor,
+        lengths: Optional[Tensor] = None,
+    ) -> Tensor:
+        x, mask = self._preprocess(features, lengths)
+        x = x.transpose(0, 1)
+        for layer in self.conformer:
+            x = layer(x, mask)
+        return x.transpose(0, 1)
+
+    def get_intermediate_outputs(
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:94]: shall we make the method private?

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+) -> Wav2Vec2Model:
+    """Build a custom Conformer Wav2Vec2Model
+
+    Args:
+        extractor_input_dim (int): Input dimension of the features.
+        extractor_output_dim (int): Output dimension after feature extraction.
+        extractor_stride (int): Stride used in time reduction layer of feature extraction.
+        encoder_embed_dim (int): The dimension of the embedding in the feature projection.
+        encoder_projection_dropout (float):
+            The dropout probability applied after the input feature is projected to ``embed_dim``
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:224]:
```suggestion
            The dropout probability applied after the input feature is projected to ``encoder_embed_dim ``
```

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 4 above)
Comment (mthrok) [reply torchaudio/prototype/models/conformer_wav2vec2.py:11]: That sounds more fitting to me as the docstring says "Extract features from input. Consists of time reduction **and** linear layer."

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        lengths: Optional[Tensor],
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        """
+        Args:
+            x (Tensor):
+                Input Tensor representing log Mel Spectrogram output. shape ``(batch, frame, dim)``.
+            length (Tensor or None, optional):
+                Valid length of each input sample. shape: ``(batch, )``.
+        """
+        # TODO: case if length is none?
```
Comment (mthrok) [inline torchaudio/prototype/models/conformer_wav2vec2.py:37]: What is supposed to happen if length is None?

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
+from typing import List, Optional, Tuple, Union
```
Comment (mthrok) [inline torchaudio/prototype/models/conformer_wav2vec2.py:1]: Could you prefix the file with underscore so as to emphasis that client code should import model from `torchaudio.[prototype.]models`?

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 4 above)
Comment (hwangjeff) [reply torchaudio/prototype/models/conformer_wav2vec2.py:11]: yeah if we're using the docstring as the reference, `FeatureExtractor` sounds good. although, if we're saying that the output of this module is features, what's the input?

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 8 above)
Comment (hwangjeff) [reply torchaudio/prototype/models/conformer_wav2vec2.py:37]: supplying some dummy tensor to `self.time_reduction_layer` should suffice

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+    def forward(
+        self,
+        x: Tensor,
+        lengths: Optional[Tensor],
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        """
+        Args:
+            x (Tensor):
+                Input Tensor representing log Mel Spectrogram output. shape ``(batch, frame, dim)``.
+            length (Tensor or None, optional):
```
Comment (hwangjeff) [inline torchaudio/prototype/models/conformer_wav2vec2.py:34]:
```suggestion
            length (Tensor or None):
```

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 8 above)
Comment (nateanl) [reply torchaudio/prototype/models/conformer_wav2vec2.py:37]: maybe initialize a lengths tensor with shape `(batch,)`, and every value is the second dimension of x.

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 10 above)
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:34]:
```suggestion
            lengths (Tensor or None):
```

File: torchaudio/prototype/models/conformer_wav2vec2.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+                Valid length of each input sample. shape: ``(batch, )``.
+        """
```
Comment (nateanl) [inline torchaudio/prototype/models/conformer_wav2vec2.py:36]: need to add docstring for returned items.

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 4 above)
Comment (carolineechen) [reply torchaudio/prototype/models/conformer_wav2vec2.py:11]: docstring was copied over from w2v2 model when I had originally named the class ConformerFeatureExtractor, and the input is waveform transforms (logmel spectrograms), so not too sure if feature extractor is fully appropriate for the name

**Review by nateanl — APPROVED** [review]

File: torchaudio/prototype/models/conformer_wav2vec2.py (hunk 4 above)
Comment (carolineechen) [reply torchaudio/prototype/models/conformer_wav2vec2.py]: renaming it to `FeatureEncoder` for now, as is used in the [paper](https://arxiv.org/pdf/2110.07313.pdf). can follow up on naming if there are any other ideas

**Review by nateanl — APPROVED** [review]

**Pull request closed without merging**